### ISSUE TITLE
Crash using sympy piecewise

### ISSUE THREAD

**Issue by dansand** [underworldcode/underworld3-archive#25]
My model crashes when I use the following viscosity expression. 

```
stokes.viscosity = Piecewise( (upperMantleViscosity, Abs(materialVariable.fn - upperMantleIndex) < 0.5),
                 (lowerMantleViscosity, Abs(materialVariable.fn -  lowerMantleIndex) < 0.5),
                 (slabViscosity, Abs(materialVariable.fn - upperSlabIndex) < 0.5),
                 (slabViscosity , Abs(materialVariable.fn - lowerSlabIndex) < 0.5),
                 (coreViscosity, Abs(materialVariable.fn - coreSlabIndex) < 0.5), 
                            ( lowerMantleViscosity,                                True ))

```

The error message says:

```
/usr/local/lib/python3.8/dist-packages/sympy/matrices/expressions/matexpr.py in doit(self, **kwargs)
    716         else:
    717             args = self.args
--> 718         return args[0][args[1], args[2]]
    719 
    720     @property

TypeError: 'Dummy' object is not subscriptable

```

Following sympy issue is the culprit (thanks @jmansour ): 

https://github.com/sympy/sympy/issues/14933#issuecomment-523197616

**Comment by jmansour** [comment]
Thanks @dansand 

I've updated the sympy ticket, as their latest changes appear to be producing incorrect results. 

In the meantime, I'll have a think about about a work around.

**Comment by julesghub** [comment]
@jmansour's new ticket https://github.com/sympy/sympy/issues/22175

**Comment by jmansour** [comment]
I've pushed up some changes in ac619ac080f8ebc5c48c2a54c500abf23be0e4d9. 

The fixes for `sympy` aren't in the master branch yet, so you'll have to install this one:

```shell
pip install git+https://github.com/smichr/sympy.git@pwoo
```

The behaviour seems correct now, although there are still some performance concerns with `Piecewise`, namedly slow equation substitutions. On my machine, it's ~12s for all the required subs for Dan's model. While we pay this price only once per module generation (ie not for subsequent solves), this delay would be quite frustrating during model development.  I have some relatively straighforward ideas about how we can significantly reduce our reliance on these substitutions, but I believe the `sympy` guys are also aware of the performance issues so hopefully there'll be some improvements on their end before it reaches the master branch. I'll plug in my changes shortly. 

However `Piecewise` mightn't be the best option for other reasons. As discussed on Friday, a _smooth_ alternative might give better mathematical behaviour. But a further and more critical issue with the Piecewise approach results due to the fact that the `materialVariable` is interpolated to determine which clause to use. This is fine when there's only two materials, but when there's more, the interface between two materials might result in a third material spuriously appearing. For example, if we have material indexes {1,2,3}, at the interface between materials {1} and {3}, the interpolated result might yield material {2}. So a mask approach is really the only correct way forward here I believe.

**Comment by julesghub** [comment]
Does this mean the code uw3 is dependent on the sympy branch `pwoo`?

I totally agree with mask idea of the materialVariable being interpolated. 
The solution strategy of masking and general smoothing of discreet vars is a concept we should tease out more and ultimately control from the python layer.

**Comment by jmansour** [comment]
Until the changes from `pwoo` are merged in to `master`, and if you want to use `sympy.Piecewise`, yes. 

But I guess my main suggestion is that we move away from `sympy.Piecewise` as the standard tool for branching.

**Comment by jmansour** [comment]
This works with `sympy==1.8` with these latest changes.  We avoid most of the expensive substitituions now, but the derivatives still take ~2s to compute, but it's a price that'll generally only be paid once, and I think without `Piecewise()` the derivs will be much faster too.

**Comment by jmansour** [comment]
I'll leave open until Dan can test.
